compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java

```diff
@@ -173,9 +174,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
         } catch (IOException e) {
             throw new RuntimeException("Error while accessing Distribution cache: " + e.getMessage());
         }
+
+        versions.removeAll(getIncompatibleVer(versions, org, name));
         return pathToVersions(versions);
     }
 
+    protected List<Path> getIncompatibleVer(List<Path> versions, PackageOrg org, PackageName name) {
+        List<Path> incompatibleVersions = new ArrayList<>();
+
+        if (!versions.isEmpty()) {
+            for (Path ver : versions) {
+                Path pkgJsonPath = getPackagePath(org.value(), name.value(),
+                        Optional.of(ver.getFileName()).get().toFile().getName()).resolve(ProjectConstants.PACKAGE_JSON);
+                if (Files.exists(pkgJsonPath)) {
+                    String packageVer = BalaFiles.readPkgJson(pkgJsonPath).getBallerinaVersion();
+                    String packVer = RepoUtils.getBallerinaShortVersion();
+                    if (!isCompatible(packageVer, packVer)) {
+                        incompatibleVersions.add(ver);
+                    }
+                } else {
+                    incompatibleVersions.add(ver);
+                }
+            }
+        }
+
+        return incompatibleVersions;
+    }
+
+    private boolean isCompatible(String pkgBalVer, String distBalVer) {
+        if (!pkgBalVer.equals(distBalVer)) {
+            String packageStage = pkgBalVer.substring(0, pkgBalVer.length() - 1);
```

Comment:
Shall we rename  this variables to be more explanatory? e.g: `pkgBalVersionPrefix`

Reply:
We can also update the comments

Reply:
Changed the isCompatible() variable names and comments